[DOC-14397]: DOC: Application Telemetry needs additional info. (#4123)
couchbase/docs-server: modules/metrics-reference/pages/application-telemetry-metrics.adoc

= Application Telemetry Metrics

> :description: Couchbase can be configured to collect metrics related to Couchbase SDK calls made from applications.

Reviewer: Add to the description that SDK's send metrics to the server.  Something like:

Couchbase Server can be configured to collect metrics related to Couchbase SDK calls made from applications.  The client SDKs send metrics to the Server that are then handled like regular server-side metrics, and you can access them through the Prometheus endpoint and the Stats API.

:stem: asciimath

[abstract]
{description}

== Prerequisites

=== Activating Telemetry on Couchbase Server

> Before generating metrics from application telemetry, you must enable application telemetry on your Couchbase Server installation. 

Reviewer: On the SDK side, "generating" app telemetry metrics is on by default and probably transparent to the customers.  So, I think that the phrasing "Before generating metrics from application telemetry" is likely to be confusing to customers -- they may think that they need to do something on the application side, even though there's really nothing they have to do (except to use an SDK version that has this feature).

So, I think that it's best to leave out the phrase "Before generating metrics from application telemetry", and just say:

You must enable application telemetry collection on your Couchbase Server installation.

You can do this by executing the following CURL command:

[source, bash, subs="+quotes"]
----
curl -u Administrator:password -X POST \

> https://##{host}##:##{port}##/settings/appTelemetry \

Reviewer: Should be:

`http[s]://{host}:{port}/settings/appTelemetry`

where the default management port is 8091 (for unencrypted connection using http) or 18091 (for encrypted connection using https)

-d enabled=true 
----

For more information on enabling application telemetry, see xref:rest-api:application-telemetry.adoc[]

== Telemetry Counters

Couchbase can record the number of SDK calls made for specific services. The metrics follow the same basic patterns:

[%header]
|===
| Metric| Description

| `sdk_\{service}_r_total`
| The total number of operations

|`sdk_\{service}_r_canceled`
| The total number of operations that were canceled.

| `sdk_\{service}_r_timedout`
| The total number of operations that timed out.

|===

To access the counter for a particular service, you replace `\{service}` with one of the following:

* `kv`
* `query`
* `search`
* `analytics`
* `management`
* `eventing`

So to access the number of calls that timed out for the query service, 
you would use the command: `sdk_query_r_timedout`.

== Histograms

> The same service labeling is used in the retrieval of histograms:

Reviewer: Actually, the statement "The same service labeling is used in the retrieval of histograms:" is not true.
Should just say:

Below are the available histograms.  The `_bucket` at the end of the metrics names is a special suffix that signifies that the line should be interpreted as a bucket of the histogram.  Also, there are two special metrics associated with each histogram: `_sum` and `_count` that represent the overall sum of all histogram buckets, and the total number of the points in the histogram.  The `_sum` value should be a total sum of the points in milliseconds represented as an integer.

[%header, cols="2,1"]
|===
| Metric| Description

| `sdk_\{service}_retrieval_duration_milliseconds_bucket`
| Time take for retrieval

|`sdk_\{service}_mutation_nondurable`
| Time taken for non-durable mutations.

| `sdk_\{service}__mutation_durable`
| Time taken for ndurable mutations.
|===

The histogram metrics can take the same `service` values as the metric counters.

* `kv`
* `query`
* `search`
* `analytics`
* `management`

> * `eventing`

Reviewer: This comment applies to all lines 60 - 81.  Should be replaced with just a list of metric names.

[%header, cols="1"]
|===
| Metric Name

| sdk_kv_retrieval_duration_milliseconds_bucket

| sdk_kv_retrieval_duration_milliseconds_count

| sdk_kv_retrieval_duration_milliseconds_sum

| sdk_kv_mutation_nondurable_duration_milliseconds_bucket

| sdk_kv_mutation_nondurable_duration_milliseconds_count

| sdk_kv_mutation_nondurable_duration_milliseconds_sum

| sdk_kv_mutation_durable_duration_milliseconds_bucket

| sdk_kv_mutation_durable_duration_milliseconds_count

| sdk_kv_mutation_durable_duration_milliseconds_sum

| sdk_query_duration_milliseconds_bucket

| sdk_query_duration_milliseconds_count

| sdk_query_duration_milliseconds_sum

| sdk_search_duration_milliseconds_bucket

| sdk_search_duration_milliseconds_count

| sdk_search_duration_milliseconds_sum

| sdk_analytics_duration_milliseconds_bucket

| sdk_analytics_duration_milliseconds_count

| sdk_analytics_duration_milliseconds_sum

| sdk_eventing_duration_milliseconds_bucket

| sdk_eventing_duration_milliseconds_count

| sdk_eventing_duration_milliseconds_sum

| sdk_management_duration_milliseconds_bucket

| sdk_management_duration_milliseconds_count

| sdk_management_duration_milliseconds_sum

|===

sdk_kv_retrieval_duration_milliseconds_bucket{le="10",agent="sdk/2.4.5.0",bucket="travel-sample",node="node1",node_uuid="91442eb8202e0e16bbb59624d9ccdb0a"} 33444
sdk_kv_retrieval_duration_milliseconds_bucket{le="100",agent="sdk/2.4.5.0",bucket="travel-sample",node="node1",node_uuid="91442eb8202e0e16bbb59624d9ccdb0a"} 100392
sdk_kv_retrieval_duration_milliseconds_bucket{le="500",agent="sdk/2.4.5.0",bucket="travel-sample",node="node1",node_uuid="91442eb8202e0e16bbb59624d9ccdb0a"} 129389
sdk_kv_retrieval_duration_milliseconds_bucket{le="1000",agent="sdk/2.4.5.0",bucket="travel-sample",node="node1",node_uuid="91442eb8202e0e16bbb59624d9ccdb0a"} 133988
sdk_kv_retrieval_duration_milliseconds_bucket{le="2500",agent="sdk/2.4.5.0",bucket="travel-sample",node="node1",node_uuid="91442eb8202e0e16bbb59624d9ccdb0a"} 139823
sdk_kv_retrieval_duration_milliseconds_bucket{le="+Inf",agent="sdk/2.4.5.0",bucket="travel-sample",node="node1",node_uuid="91442eb8202e0e16bbb59624d9ccdb0a"} 144320
sdk_kv_retrieval_duration_milliseconds_sum{agent="sdk/2.4.5.0",bucket="travel-sample",node="node1",node_uuid="91442eb8202e0e16bbb59624d9ccdb0a"} 53423000

> sdk_kv_retrieval_duration_milliseconds_count{agent="sdk/2.4.5.0",bucket="travel-sample",node="node1",node_uuid="91442eb8202e0e16bbb59624d9ccdb0a"} 144320 

Reviewer: In the Server,  not all of the labels seen are kept.  
Below is like what you would see if you ran 
curl -X GET -u Administrator:password http://localhost:8091/metrics | grep -i '^sdk' | sort
on your cluster after you've enabled application telemetry on the Server (and ran an application that did queries and stayed connected to the cluster longer than the Server scrape interval for application telemetry -- by default, 60 seconds)

So should just show:

sdk_kv_retrieval_duration_milliseconds_bucket{le="1"} 24054
sdk_kv_retrieval_duration_milliseconds_bucket{le="10"} 33444
sdk_kv_retrieval_duration_milliseconds_bucket{le="100"} 100392
sdk_kv_retrieval_duration_milliseconds_bucket{le="500"} 129389
sdk_kv_retrieval_duration_milliseconds_bucket{le="1000"} 133988
sdk_kv_retrieval_duration_milliseconds_bucket{le="2500"} 139823
sdk_kv_retrieval_duration_milliseconds_bucket{le="+Inf"} 144320
sdk_kv_retrieval_duration_milliseconds_sum{} 53423000
sdk_kv_retrieval_duration_milliseconds_count{} 144320

----
====
xref:histogram-output[xrefstyle=short] demonstrates a series of `kv` retrieval times for the buckets in a node.
Each line gives the retrieval count for a particular range. 

> The retrieval duration counts are grouped by latency buckets: for example, requests completed in stem:[le 1] ms total 24,054, those in stem:[le 10] ms total 33,444, and so on.

Reviewer: Each line gives the retrieval count for a particular range.
For example, the `le="1"` label is the upper bound of the latency bucket, so that range is <=1ms.  The value would fall into all buckets where upper bound is higher.
As can be seen, for the retrieval duration values, the requests completed in stem:[le 1] ms total 24,054, those in stem:[le 10] ms total 33,444, and so on.

The `_bucket` at the end of the metric names is a special suffix that signifies that the line should be interpreted as a bucket of that histogram. 
Also, there are 2 special metrics associated with each histogram: `_sum` and `_count` that represent the overall sum of all histogram buckets, 
and the total number of the points in the histogram. 
The `_sum` value should be a total sum of the points in milliseconds represented as an integer.

[#available-histograms]
=== Available Histograms

* `sdk_kv_retrieval_duration_milliseconds_bucket`
* `sdk_kv_retrieval_duration_milliseconds_count`
* `sdk_kv_retrieval_duration_milliseconds_sum`
* `sdk_kv_mutation_nondurable_duration_milliseconds_bucket`
* `sdk_kv_mutation_nondurable_duration_milliseconds_count`
* `sdk_kv_mutation_nondurable_duration_milliseconds_sum`
* `sdk_kv_mutation_durable_duration_milliseconds_bucket`
* `sdk_kv_mutation_durable_duration_milliseconds_count`
* `sdk_kv_mutation_durable_duration_milliseconds_sum`
* `sdk_query_duration_milliseconds_bucket`
* `sdk_query_duration_milliseconds_count`
* `sdk_query_duration_milliseconds_sum`
* `sdk_search_duration_milliseconds_bucket`
* `sdk_query_duration_milliseconds_count`

> * `sdk_query_duration_milliseconds_sum`

Reviewer: Looks like lines 82-84 for copy/paste duplicates -- should be deleted.

Author: Yes, you're right! Good spot. I completely missed that.

[%header]
|===
| Metric| Description

| `sdk_\{service}_r_total`
| The total number of operations

|`sdk_\{service}_r_canceled`
| The total number of operations that were canceled.

| `sdk_\{service}_r_timedout`
| The total number of operations that timed out.

|===

To access the counter for a particular service, you replace `\{service}` with one of the following:

* `kv`
* `query`
* `search`
* `analytics`
* `management`
* `eventing`

So to access the number of calls that timed out for the query service, 
you would use the command: `sdk_query_r_timedout`.

== Histograms

The same service labeling is used in the retrieval of histograms:

[%header, cols="2,1"]
|===
| Metric| Description

| `sdk_\{service}_retrieval_duration_milliseconds_bucket`
| Time take for retrieval

|`sdk_\{service}_mutation_nondurable`
| Time taken for non-durable mutations.

| `sdk_\{service}__mutation_durable`
| Time taken for ndurable mutations.
|===

The histogram metrics can take the same `service` values as the metric counters.

* `kv`
* `query`
* `search`
* `analytics`
* `management`
* `eventing`

[#histogram-output]
.Histogram output

> ====

Reviewer 2: ```suggestion
[source]
----
```

Author: Leaving this because I want to have the example to be numbered.

> #sdk_kv_retrieval_duration_milliseconds_bucket{le="1",agent="sdk/2.4.5.0",bucket="travel-sample",node="node1",node_uuid="91442eb8202e0e16bbb59624d9ccdb0a"} 24054#

Reviewer 2: ```suggestion
sdk_kv_retrieval_duration_milliseconds_bucket{le="1",agent="sdk/2.4.5.0",bucket="travel-sample",node="node1",node_uuid="91442eb8202e0e16bbb59624d9ccdb0a"} 24054
```

sdk_kv_retrieval_duration_milliseconds_bucket{le="100",agent="sdk/2.4.5.0",bucket="travel-sample",node="node1",node_uuid="91442eb8202e0e16bbb59624d9ccdb0a"} 100392
sdk_kv_retrieval_duration_milliseconds_bucket{le="500",agent="sdk/2.4.5.0",bucket="travel-sample",node="node1",node_uuid="91442eb8202e0e16bbb59624d9ccdb0a"} 129389
sdk_kv_retrieval_duration_milliseconds_bucket{le="1000",agent="sdk/2.4.5.0",bucket="travel-sample",node="node1",node_uuid="91442eb8202e0e16bbb59624d9ccdb0a"} 133988
sdk_kv_retrieval_duration_milliseconds_bucket{le="2500",agent="sdk/2.4.5.0",bucket="travel-sample",node="node1",node_uuid="91442eb8202e0e16bbb59624d9ccdb0a"} 139823
sdk_kv_retrieval_duration_milliseconds_bucket{le="+Inf",agent="sdk/2.4.5.0",bucket="travel-sample",node="node1",node_uuid="91442eb8202e0e16bbb59624d9ccdb0a"} 144320
sdk_kv_retrieval_duration_milliseconds_sum{agent="sdk/2.4.5.0",bucket="travel-sample",node="node1",node_uuid="91442eb8202e0e16bbb59624d9ccdb0a"} 53423000
sdk_kv_retrieval_duration_milliseconds_count{agent="sdk/2.4.5.0",bucket="travel-sample",node="node1",node_uuid="91442eb8202e0e16bbb59624d9ccdb0a"} 144320 

> ====

Reviewer 2: ```suggestion
----
```

Author: See above.

<<histogram-output>> demonstrates a series of `kv` retrieval times for the buckets in a node.
Each line gives the retrieval count for a particular range. 

> (The count for retrieval durations for milliseconds stem:[le 1] is 24,054, milliseconds stem:[le 10] is 33,444, etc).

Reviewer 2: ```suggestion
The retrieval duration counts are grouped by latency buckets: for example, requests completed in `<= 1 ms` total `24,054`, those in `<=10 ms` total `33,444`, and so on.
```
this paragraph is not clear. I believe "bucket" term will be better fit here.

Author: Leaving the STEM because AsciiDoc changes the symbol to a double-lined arrow.
